lib/puppet/type/k5login.rb

```diff
@@ -0,0 +1,164 @@
+      end
+    end
+  end
+
+  # To manage the mode of the file
+  newproperty(:mode) do
+    desc 'The desired permissions mode of the `.k5login` file. Defaults to `644`.'
+    defaultto { '644' }
+  end
+
+  # To manage the selinux user of the file
+  newproperty(:seluser, parent: Puppet::SELFileContext) do
+    desc "What the SELinux user component of the context of the file should be.
+      Any valid SELinux user component is accepted.  For example `user_u`.
+      If not specified it defaults to the value returned by matchpathcon for
+      the file, if any exists.  Only valid on systems with SELinux support
+      enabled."
+
+    defaultto { 'user_u' }
+  end
+
+  # To manage the selinux role of the file
+  newproperty(:selrole, parent: Puppet::SELFileContext) do
+    desc "What the SELinux role component of the context of the file should be.
+      Any valid SELinux role component is accepted.  For example `role_r`.
+      If not specified it defaults to the value returned by matchpathcon for
+      the file, if any exists.  Only valid on systems with SELinux support
+      enabled."
+
+    defaultto { 'object_r' }
+  end
+
+  # To manage the selinux type of the file
+  newproperty(:seltype, parent: Puppet::SELFileContext) do
+    desc "What the SELinux type component of the context of the file should be.
+      Any valid SELinux type component is accepted.  For example `tmp_t`.
+      If not specified it defaults to the value returned by matchpathcon for
+      the file, if any exists.  Only valid on systems with SELinux support
+      enabled."
+
+    # to my knowledge, `krb5_home_t` is the only valid type for .k5login
+    defaultto { 'krb5_home_t' }
+  end
+
+  # To manage the selinux range of the file
+  newproperty(:selrange, parent: Puppet::SELFileContext) do
+    desc "What the SELinux range component of the context of the file should be.
+      Any valid SELinux range component is accepted.  For example `s0` or
+      `SystemHigh`.  If not specified it defaults to the value returned by
+      matchpathcon for the file, if any exists.  Only valid on systems with
+      SELinux support enabled and that have support for MCS (Multi-Category
+      Security)."
+
+    defaultto { 's0' }
+  end
+
+  # Stat our file.
+  #
+  # We use the initial value :needs_stat to ensure we only stat the file once,
+  # but can also keep track of a failed stat (@stat == nil). This also allows
+  # us to re-stat on demand by setting @stat = :needs_stat.
+  def stat
+    return @stat unless @stat == :needs_stat
+
+    @stat = begin
+      Puppet::FileSystem.stat(self[:path])
+    rescue Errno::ENOENT
+      nil
+    rescue Errno::ENOTDIR
+      nil
+    rescue Errno::EACCES
+      warning _('Could not stat; permission denied')
+      nil
+    end
+  end
+
+  def initialize(args)
+    @stat = :needs_stat
+    super
+  end
+
+  # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+  provide(:k5login) do
```

Comment:
@joshcooper @jhelwig should I break this update into seperate files, or leave it as is?

Reply:
oh ha, there's the provider! I think it's better to move this into a separate file, because puppet creates different autoloader instances for types vs providers. As it is now, the k5login provider will be autoloaded by the type autoloader. It also means we always load the provider on the master, which we try to avoid.

Reply:
@joshcooper Seems like it might be worth deprecating the `provide` method on Type to discourage people (including us) from combining the Type & Provider like this?